Conflict Identification

Deco354 opened the pull request:
Part of [Conflict Resolution Dialog Issue](https://github.com/wordpress-mobile/WordPress-iOS/issues/14798)

This PR fixes bugs where posts with local autosaves are either not updated or disappear and adds a version conflict label to them so users can see there's a version conflict

A later PR will cover the resolution of this conflict.

## Fixes 
* Cells within post list now display "Version conflict" label if they have a conflicting local autosave and 
* Locally autosaved drafts that are published remotely remotely no longer disappear
* Autosaved published posts that are updated remotely no longer ignore remote updates

## Basic Test cases
### 1. Local Autosave Draft vs Remote Update
#### Reproduction Steps
* Create a post using the app and save as a draft
* Open the post, type some content and then close the app (without exiting the editor)
* Go to the web, change the post content and publish it

#### Desired outcomes:
- The post does not appear in the drafts tab
- The post appears within the published tab
- The post has a Version Conflict label

### 2. Local Autosave Publication vs Remote Update
#### Reproduction Steps
* Create a post on the web and publish
* Open the post in the app, type some content and then close the app (without exiting the editor)
* Go to the web, change the post content and publish the changes.

#### Desired Outcome
- The post appears in the published tab
- The post has a Version Conflict label

## Known issues (WIP)
* False positive version conflict will sometimes appear without a remote autosave or update being applied as the date modified can change without either happening.
* Work to unify labels for remote and local version conflicts may be reverted seeing as the outcomes of these conflicts are very different.

PR submission checklist:

- [ ] I have considered adding unit tests where possible.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Deco354 committed:
Add hasVersionConflict to AbstractPost

Includes necessary core data update as Abstract Post is
a descendant of NSManagedObject.

Deco354 committed:
Flag conflicting posts and their revisions

Flags posts with remote updates and local revisions (autosaves)
as conflicting so we can handle this later.

Deco354 committed:
Fix drafts with revisions disappearing upon update

Changes predicate to allow original posts with conflicts
and revisions in other tabs to be displayed.

This can happen when a local draft has been autosaved then that same post
is published remotely. The published post is not a revision but
should be displayed as the revision is in the draft tab.

Deco354 committed:
Display versionConflict within PostCell

Deco354 committed:
Use version conflict label for remote autosaves

Now we account for two types of version conflict we can utilize the
same design for them both and differentiate between them upon displaying
the version conflict prompt

Deco354 committed:
Move duplication predicate to private function

Deco354 committed:
Remove published posts from drafts

If a draft with a local autosave has been published remotely it will
no longer appear within the draft tab.

Deco354 committed:
Merge branch 'develop' of https://github.com/Deco354/WordPress-iOS into feature/conflict-identification

# Conflicts:
#	WordPress/Classes/WordPress.xcdatamodeld/WordPress 99.xcdatamodel/contents
#	WordPress/WordPress.xcodeproj/project.pbxproj

diegoreymendez commented on WordPress/WordPressTest/PostCardCellTests.swift:
```diff
@@ -490,13 +490,22 @@ class PostCardCellTests: XCTestCase {
 
-        expect(self.postCell.statusLabel.text).to(equal(i18n("You've made unsaved changes to this post")))
-        expect(self.postCell.statusLabel.textColor).to(equal(UIColor.warning(.shade40)))
+        expect(self.postCell.statusLabel.text).to(equal(i18n("Version Conflict")))
```
I understand the test below showing a version conflict as the post is constructed with hasVersionConflict(). This one though isn't... and I'm failing to understand the logic.

Can you explain a bit why the version conflict is correct in this test?

Deco354 replied:
This test checks that the local autosave conflict label has replaced the remote autosave conflict label for remote conflicts. This was back when we intended to use the same dialog for both local and remote conflicts which after seeing that they exhibit very different behavior and speaking with yourself we decided against.

Perhaps it does make sense to have different labels for remote and local autosave conflicts now that we're using two distinct dialogs. What do you think? It will be easy enough to revert this commit and go back to having the old style label for remote auto saves conflicts.

#### Current Remote autosave conflict label
![image](https://user-images.githubusercontent.com/18557720/95489912-5ff81800-098f-11eb-887a-d554832aad1d.png)

#### New local autosave conflict label
![image](https://user-images.githubusercontent.com/18557720/95490442-24aa1900-0990-11eb-93fe-43a00639860a.png)

Cliffs: Do we want to keep different labels for these both?

diegoreymendez replied:
As chatted via DM, I think the previous label was better for autosave conflicts.  We can re-evaluate the UI once we're finished though.

diegoreymendez commented on WordPress/Classes/Services/PostService.m:
```diff
@@ -786,6 +786,8 @@ - (void)updatePost:(AbstractPost *)post withRemotePost:(RemotePost *)remotePost
     post.content = remotePost.content;
     post.status = remotePost.status;
     post.password = remotePost.password;
+    post.hasVersionConflict = post.hasRevision;
```
This looks off. Having a revision shouldn't be enough for a post to have a version conflict. I believe we discussed this at some point, and I also know this is a work in progress.

Can you clarify if this is changing in an upcoming PR?

Deco354 replied:
Very true. This is where the `remoteContentHash` you suggested will be used to confirm that an update with new content. has been fetched. If a remote update with new content that is editable within a local revision arrive then we'll mark it as a conflict.

I previously had an if statement here comparing the `dateModified` field of the remote and local post although unfortunately that doesn't work as the dateModified field is prone to change when updates are made even when no save or autosave takes place so I dropped it from my last rebase seeing as we'll be replacing it anyway.

![image](https://user-images.githubusercontent.com/18557720/95488144-160e3280-098d-11eb-9a04-40decbcafc98.png)

I think I'll put the hash storage and utilization of the hash within this PR seeing as I'll need to do the actual hash implementation separately within WordPressKit.

diegoreymendez replied:
We can tackle this in the next PR.

diegoreymendez reviewed (commented):
Hey looking good. I added some questions for you.

diegoreymendez reviewed (approved):
I'm approving the PR.  Feel free to address the label change in this PR or the next one.

None of the two issues mentioned should be blockers to merge this since we're working in a feature branch.

Deco354 committed:
Revert "Use version conflict label for remote autosaves"

This reverts commit 7f40fbb29ea4dc79a39f25fdaa003dcb7989c641.

Deco354 committed:
Resolve version conflict false positives

Only flag version conflicts if the remote posts modification
date differs from the local post.
